JavaScript_Weekly/571/fuite.md

```diff
@@ -0,0 +1,95 @@
+
+因此我写了一个工具`fuite (法语“ leak”)`。在这个CLI工具中，我们可以用通过它来分析任意的一个网址是否存在内存泄漏的问题。
+```
+npx fuite https://example.com
+```
+在执行命令时，`fuite`会默认网站是呈现给用户的一个SPA页面，之后会抓取页面中的内部链接(例如``/about ``或``/contact``)。获取到链接后，会按照下面步骤来执行：
+1. 点击链接
+2. 点击浏览器的返回按钮
+3. 重复操作来监测内存是否增长
+当`fuite`监测到内存异常时，会告知我们哪些对象可能会导致内存泄漏。
+```
+Test         : Go to /foo and back
+Memory change: +10 MB
+Leak detected: Yes
+Leaking objects:
+| Object            | # added | Retained size increase |
+| ----------------- | ------- | ---------------------- |
+| HTMLIFrameElement | 1       | +10 MB                 |
+Leaking event listeners:
+| Event        | # added | Nodes  |
+| ------------ | ------- | ------ |
+| beforeunload | 2       | Window |
+Leaking DOM nodes:
+DOM size grew by 6 node(s) 
+```
+为了实现这个功能，我在`fuite`中使用了博客文章中阐述的基本策略。他会启动Chrome来运行n次(默认7次),并检查是否有对象发生内存泄漏了n次(7,14,21等）。
+
+`fuite`同时会分析全部的 ``Array``, ``Object``,`` Map``, ``Set``，事件监听器以及整个DOM。假如一个数组在执行了7次后同时也增长了7次，那么这个对象就有可能发生了内存泄漏。
+## 测试正式网站
+在反复进行页面跳转并返回的操作时，会发现在许多spa应用中发现内存泄漏的问题。下面是我对10个前端主流框架的首页进行的测试结果,并发现他们都存在内存泄漏的问题。
+
+| 站点    | 是否发生泄漏 | 内部连接数 | 平均增长 | 最大增长数 |
+| ------- | ------------- | -------------- | -------------- | ---------- |
+| Site 1  | yes           | 8              | 27.2 kB        | 43 kB      |
+| Site 2  | yes           | 10             | 50.4 kB        | 78.9 kB    |
+| Site 3  | yes           | 27             | 98.8 kB        | 135 kB     |
+| Site 4  | yes           | 8              | 180 kB         | 212 kB     |
+| Site 5  | yes           | 13             | 266 kB         | 1.07 MB    |
+| Site 6  | yes           | 8              | 638 kB         | 1.15 MB    |
+| Site 7  | yes           | 7              | 1.37 MB        | 2.25 MB    |
+| Site 8  | yes           | 15             | 3.49 MB        | 4.28 MB    |
+| Site 9  | yes           | 43             | 5.57 MB        | 7.37 MB    |
+| Site 10 | yes           | 16             | 14.9 MB        | 186 MB     |
+
+在这种情况下，`内部链接`是指所测试的内部链接数目，`平均增长`是指每个链接的平均内存增长(即每次跳转页面并返回) ，而`最大增长`是指内部链接内存泄漏的最大值。值得注意的时，这些数据并不是一次的测试结果而是`fuite`测试了7次之后的结果。
+
+当我们要确定这些结果的真实性时,你可以使用Chrome DevTools的`Memory`工具。下面图片是我选择了测试中表现最差的站点，点击其中的一个链接，然后按下后退键。每次操作都保存一个堆得快照，然后重复操作。
+
+![74a805399e214e6a054c93ad6e9d4ba.jpg](https://p9-juejin.byteimg.com/tos-cn-i-k3u1fbpfcp/34e6836feb8d47018b63611dd73ae2ff~tplv-k3u1fbpfcp-watermark.image?)
+
+我们可以发现每次当点击页面的时候，内存大概都会增加6M左右。
+## 注意事项
+我们要注意的是并不是所有的SPA中的内存泄漏都是需要解决的。比如，当SPA需要维护焦点和滚动状态，虽然会造成泄漏，但是却并不致命。`fuite`会提示我们这些小的问题，因为它确实导致了内存泄漏，但是作为开发者我们需要注意这种问题是否真的需要修改。
+
+在开发过程中，一些内存的增长可能是由于浏览器内部的原因造成的(比如JITing)，这是网页无法真正控制的。正因如此，内存增长并不意味着你的代码存在问题，很多情况下内存的增长是不可避免的。
+
+在少数情况下，由于浏览器的一些bug也会造成内存的增长。在分析上面网站时，由于Chrome的bug导致站点#4出现了问题。由于`fuite`是程序，所以很难检测到是否是浏览器的问题。在`fuite`有警告的情况下，我们可以对比下其他浏览器，来确定问题究竟来源于何处。
+
+说完了SPA的应用，在我们传统的MPA(多页面应用程序)中，就很少会出现内存泄漏的问题，因为每次当我们切换页面时，都会清除浏览器的内存。当然，这都是在浏览器没有bug的前提下。
```

Comment:
Also note that it’s almost impossible for a Multi-Page App (MPA) to leak, because the browser clears memory on every page navigation. (Assuming no browser bugs, of course.) During my testing, I found two frontend frameworks whose home pages were MPAs, and unsurprisingly, fuite couldn’t find any leaks in them. These were excluded from the results above.
```suggestion
说完了SPA的应用，在我们传统的MPA(多页面应用程序)中，几乎不可能出现泄漏问题，因为每次当我们切换页面时，浏览器都会清除内存。当然，这都是在浏览器没有bug的前提下。
```